File: scripts/faces/src/clustering.py
"""Port of web/faceEmbed/clustering.ts to Python.

Same two failure modes the TS version guards against:

1. The raw mean of unit-norm embeddings has norm < 1 and shrinks as
   members diversify — distance comparisons compress and one cluster
   absorbs everything vaguely face-like. Fix: re-normalise the
   centroid to the unit sphere every time it's used.

2. The mean has no membership check after the fact; an embedding can
   join via mean drift even when it disagrees with every actual
   member. Fix: prune() finds the medoid (real member closest to the
   centroid) and drops members whose distance to the medoid is >=
   threshold, then recomputes sum/count from survivors.

The cluster representative for downstream use is the medoid — never
the centroid. SAME_CHARACTER_THRESHOLD = 1.1 (matches
web/faceEmbed/clustering.ts).
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable, Generic, List, Optional, TypeVar

import numpy as np

T = TypeVar("T")

SAME_CHARACTER_THRESHOLD = 1.1


def l2_distance(a: np.ndarray, b: np.ndarray) -> float:
    return float(np.linalg.norm(a - b))


@dataclass
class Cluster(Generic[T]):
    members: List[T] = field(default_factory=list)
    # Sum of embeddings (un-normalised) so adding a new member is O(d).
    sum: np.ndarray = field(default_factory=lambda: np.zeros(0, dtype=np.float32))
    count: int = 0

    def centroid(self) -> np.ndarray:
        """Unit-normalised mean — re-normalise EVERY call so distance
        comparisons stay metric-consistent regardless of how many
        members have been added since last call."""
        mean = self.sum / max(self.count, 1)
        norm = float(np.linalg.norm(mean))
        if norm < 1e-12:
            return mean
        return mean / norm


class OnlineClusterer(Generic[T]):
    """Single-pass clustering — for each item, find the closest existing
    cluster centroid and either join it (distance < threshold) or start
    a new one. Order matters: callers that care about determinism should
    feed items in a stable order (we feed by timestamp, then face idx)."""

    def __init__(
        self,
        threshold: float,
        get_embedding: Callable[[T], np.ndarray],
    ) -> None:
        self.threshold = threshold
        self.get_embedding = get_embedding
        self.clusters: List[Cluster[T]] = []

    def add(self, item: T) -> None:
        emb = self.get_embedding(item)
        if not self.clusters:
            self._open_cluster(item, emb)
            return
        # Find the closest existing centroid.
        best_idx = -1
        best_d = float("inf")
        for i, c in enumerate(self.clusters):
            d = l2_distance(c.centroid(), emb)
            if d < best_d:
                best_d = d
                best_idx = i
        if best_d < self.threshold and best_idx >= 0:
            c = self.clusters[best_idx]
            c.members.append(item)
            c.sum += emb
            c.count += 1
        else:
            self._open_cluster(item, emb)

    def _open_cluster(self, item: T, emb: np.ndarray) -> None:
        c: Cluster[T] = Cluster()
        c.members.append(item)
        c.sum = emb.astype(np.float32).copy()
        c.count = 1
        self.clusters.append(c)

    def prune(self) -> None:
        """Drop members whose distance to the cluster medoid is >=
        threshold, recompute sum/count from survivors. Empty clusters
        are removed."""
        kept: List[Cluster[T]] = []
        for c in self.clusters:
            if not c.members:
                continue
            centroid = c.centroid()
            medoid_idx = _argmin(
                range(len(c.members)),
                lambda i: l2_distance(self.get_embedding(c.members[i]), centroid),
            )
            medoid_emb = self.get_embedding(c.members[medoid_idx])
            survivors = [
                m for m in c.members
                if l2_distance(self.get_embedding(m), medoid_emb) < self.threshold
            ]
            if not survivors:
                continue
            new_sum = np.zeros_like(survivors[0] if isinstance(survivors[0], np.ndarray) else self.get_embedding(survivors[0]), dtype=np.float32)
            for m in survivors:
                new_sum += self.get_embedding(m)
            nc: Cluster[T] = Cluster()
            nc.members = survivors
            nc.sum = new_sum
            nc.count = len(survivors)
            kept.append(nc)
        self.clusters = kept


def _argmin(seq, key):
    seq = list(seq)
    best = seq[0]
    bestv = key(best)
    for x in seq[1:]:
        v = key(x)
        if v < bestv:
            best = x
            bestv = v
    return best


def medoid_of(items: List[T], get_embedding: Callable[[T], np.ndarray]) -> T:
    """Return the item whose embedding is closest to the running
    centroid of the set. Cluster representatives use this to pick the
    'best face' (the real face whose embedding is most central)."""
    if not items:
        raise RuntimeError("medoid_of on empty list")
    emb_stack = np.stack([get_embedding(i) for i in items])
    centroid = emb_stack.mean(axis=0)
    cn = float(np.linalg.norm(centroid))
    if cn > 1e-12:
        centroid = centroid / cn
    dists = np.linalg.norm(emb_stack - centroid[None, :], axis=1)
    return items[int(np.argmin(dists))]


def cluster_embeddings(
    items: List[T],
    threshold: float,
    get_embedding: Callable[[T], np.ndarray],
) -> List[Cluster[T]]:
    """Convenience wrapper: run an OnlineClusterer through every item
    then prune. Use the class directly if you need to interleave adds
    and inspections."""
    oc: OnlineClusterer[T] = OnlineClusterer(threshold, get_embedding)
    for it in items:
        oc.add(it)
    oc.prune()
    return oc.clusters

File: web/faceEmbed/clustering.ts
// Port of facegrabs/py/clustering.py — greedy single-pass online cluster
// assignment + medoid prune. Distance metric matches arcface.l2Distance
// (Euclidean over unit-length embeddings). Threshold 1.1 matches the
// Python pipeline's same-character threshold.

import { l2Distance } from "./arcface";

export const SAME_CHARACTER_THRESHOLD = 1.1;

export interface Cluster<T> {
    // Running sum of unit-length embeddings + count. centroid() = sum/count
    // re-normalised. Kept as the running sum so we can update cheaply.
    sum: Float32Array;
    count: number;
    members: T[];
}

function centroidOf<T>(c: Cluster<T>): Float32Array {
    const out = new Float32Array(c.sum.length);
    if (c.count === 0) return out;
    let sumSq = 0;
    for (let i = 0; i < out.length; i++) {
        out[i] = c.sum[i] / c.count;
        sumSq += out[i] * out[i];
    }
    const norm = Math.sqrt(sumSq) + 1e-12;
    for (let i = 0; i < out.length; i++) out[i] /= norm;
    return out;
}

function closestTo<T>(items: T[], reference: Float32Array, getEmbedding: (t: T) => Float32Array): T {
    let best = items[0];
    let bestD = l2Distance(getEmbedding(best), reference);
    for (let i = 1; i < items.length; i++) {
        const d = l2Distance(getEmbedding(items[i]), reference);
        if (d < bestD) { best = items[i]; bestD = d; }
    }
    return best;
}

// Medoid: the actual member whose embedding is closest to the centroid.
// Useful as a "best representative" face — the centroid itself is a
// synthetic average, never a real face crop.
export function medoidOf<T>(items: T[], getEmbedding: (t: T) => Float32Array): T {
    if (items.length === 0) throw new Error("medoidOf: empty input");
    // Build a centroid from items directly (not via Cluster wrapper).
    const dim = getEmbedding(items[0]).length;
    const sum = new Float32Array(dim);
    for (const it of items) {
        const e = getEmbedding(it);
        for (let i = 0; i < dim; i++) sum[i] += e[i];
    }
    const mean = new Float32Array(dim);
    let sumSq = 0;
    for (let i = 0; i < dim; i++) {
        mean[i] = sum[i] / items.length;
        sumSq += mean[i] * mean[i];
    }
    const norm = Math.sqrt(sumSq) + 1e-12;
    for (let i = 0; i < dim; i++) mean[i] /= norm;
    return closestTo(items, mean, getEmbedding);
}

// Online single-pass clusterer — for each new embedding, find the closest
// existing cluster by centroid distance. If within threshold, append.
// Otherwise seed a new cluster.
export class OnlineClusterer<T> {
    clusters: Cluster<T>[] = [];
    constructor(public threshold: number) { }

    assign(embedding: Float32Array): Cluster<T> {
        let best: Cluster<T> | undefined;
        let bestD = Infinity;
        for (const c of this.clusters) {
            const d = l2Distance(embedding, centroidOf(c));
            if (d < bestD) { bestD = d; best = c; }
        }
        if (best && bestD < this.threshold) {
            for (let i = 0; i < embedding.length; i++) best.sum[i] += embedding[i];
            best.count++;
            return best;
        }
        const fresh: Cluster<T> = { sum: new Float32Array(embedding), count: 1, members: [] };
        this.clusters.push(fresh);
        return fresh;
    }

    // After all assignment passes are done, drop members whose embedding
    // drifted away from the cluster's medoid by >= threshold. Recomputes
    // sum/count from survivors so the centroid no longer drags toward
    // the drift.
    prune(getEmbedding: (t: T) => Float32Array): void {
        for (const c of this.clusters) {
            if (c.members.length === 0) continue;
            const medoid = medoidOf(c.members, getEmbedding);
            const medoidEmb = getEmbedding(medoid);
            const survivors = c.members.filter(m =>
                l2Distance(getEmbedding(m), medoidEmb) < this.threshold);
            c.members = survivors;
            const dim = c.sum.length;
            const newSum = new Float32Array(dim);
            for (const m of survivors) {
                const e = getEmbedding(m);
                for (let i = 0; i < dim; i++) newSum[i] += e[i];
            }
            c.sum = newSum;
            c.count = survivors.length;
        }
    }
}

// Final consolidation pass. The online assigner is greedy and order-
// dependent, so the same character can seed two clusters (e.g. a few early
// frames drift the first centroid, then a fresh cluster forms that would
// have matched the settled centroid). This unions any pair of clusters whose
// centroids are within `threshold`, recomputing the merged centroid and
// iterating to a fixed point. O(clusters²) per pass — fine for ≤30 clusters.
export function mergeClusters<T>(clusters: Cluster<T>[], threshold: number): Cluster<T>[] {
    const out = clusters.filter(c => c.members.length > 0);
    let changed = true;
    while (changed) {
        changed = false;
        for (let a = 0; a < out.length && !changed; a++) {
            for (let b = a + 1; b < out.length; b++) {
                if (l2Distance(centroidOf(out[a]), centroidOf(out[b])) < threshold) {
                    const A = out[a], B = out[b];
                    for (let i = 0; i < A.sum.length; i++) A.sum[i] += B.sum[i];
                    A.count += B.count;
                    for (const m of B.members) A.members.push(m);
                    out.splice(b, 1);
                    changed = true;
                    break;
                }
            }
        }
    }
    return out;
}

export function clusterEmbeddings<T>(
    items: T[],
    threshold: number,
    getEmbedding: (t: T) => Float32Array,
): Cluster<T>[] {
    const cl = new OnlineClusterer<T>(threshold);
    for (const it of items) {
        const c = cl.assign(getEmbedding(it));
        c.members.push(it);
    }
    cl.prune(getEmbedding);
    return mergeClusters(cl.clusters, threshold);
}

export { l2Distance };
